### ISSUE TITLE
Memory leak. Android application process is never finished.

### ISSUE THREAD

**Issue by kamilklyta** [aws-amplify/amplify-flutter#2736]
### Description

The process of an android application is never terminated/closed if the application has isolates running and has been closed during S3 file upload.

For more details, see `Steps to Reproduce` section.

~Additional running isolates are the crux of the problem here. If they are closed/not opened when closing the application, the application process will end correctly~.

On my Pixel 6 device, the app stays in the background until the device is restarted, during this time the isolates stay operational (eg. `Timer.periodic` runs, constantly writing data to the console).

It has not yet been tested on iOS, but I can provide more details soon.

### Categories

- [ ] Analytics
- [ ] API (REST)
- [ ] API (GraphQL)
- [ ] Auth
- [ ] Authenticator
- [ ] DataStore
- [X] Storage

### Steps to Reproduce

1. Open the Flutetr app.
2. Starts a few isolates by calling the `Isolate.spawn` method. Do not close/pause them. Let's assume that they are there waiting for events from the main thread to process.
3. Start the file upload via the [`amplify_storage_s3` package, version 0.6.12](https://pub.dev/packages/amplify_storage_s3).
4. Close the app (from the UI/User perspective, swipe up, etc) during the file upload process.
5. File transfer is done in the background, the main flutter/dart isolate is closed, and the Flutter engine is detached. The isolates and the Android application process are never closed.

### Platforms

- [ ] iOS
- [X] Android
- [ ] Web
- [ ] macOS
- [ ] Windows
- [ ] Linux

### Flutter Version

3.7.6

### Amplify Flutter Version

0.6.12

### Deployment Method

Amplify CLI

**Comment by kamilklyta** [comment]
### Update: 
When there are no additional isolates, the application process **also** remains alive in the background.

However, without additional isolates running, it is possible (in my case) to run the application without error logs from other native packages that report that they are already "open".

**Comment by kamilklyta** [comment]
### Update:
As proof, since the app is not properly closed, we achieved an almost 600% increase in the "user engagement" metric, just after the app update that introduced the "S3 file upload" feature. So the other plugins, like the firebase, also remain in the memory.

___
  
I mean... Thanks, but probably there are better ways to achieve this, right? 😅.

  
<img width="634" alt="Screenshot 2023-03-08 at 17 30 03" src="https://user-images.githubusercontent.com/40759300/223773607-5ef17195-e824-4bd4-b9ec-741afcd6f50e.png">

**Comment by fjnoyp** [comment]
Hi @kamilklyta thanks for reaching out about this issue and sorry that your team is experiencing this. 

I believe I have been able to reproduce your issue.  I am sharing the details here to make sure we have the same issue.  

Can you confirm that this matches your issue? 

On an Android Pixel Emulator I start Amplify's Storage Example App. 

> I upload an image, wait for it to upload, then close the app by swiping up. 
> I go to settings -> Amplify Storage Example App -> App Info 
> I see that `Force stop` is not possible as the app is closed. 
> 
> I upload a large image, I immediately close the app by swiping up. 
> I go to settings -> Amplify Storage Example App -> App Info 
> I see that `Force stop` is **possible** because the app is not closed. 
> 

In the logs I see the following error message multiple times: 

`W/FlutterJNI(28194): Tried to send a platform message to Flutter, but FlutterJNI was detached from native C++. Could not send. Channel: com.amazonaws.amplify/storage_transfer_progress_events. Response ID: xx`


I did not use isolates as it doesn't seem like the primary cause.  I think we are improperly handling a message sent from Android to Flutter that is causing the Android process to not terminate.  It seems isolates might be retained as a secondary consequence. 


Can you clarify what you mean here? 

> However, without additional isolates running, it is possible (in my case) to run the application without error logs from other native packages that report that they are "already opened".

Are you saying not having isolates makes it harder for you to know if the app was actually terminated successfully?

**Comment by kamilklyta** [comment]
Hi @fjnoyp! Thanks for your quick action, I really appreciate it! 

I think this is the same problem.

At the time of writing the problem I was still in the process of debugging, so my first thought was that it might be caused by running isolates.

I thought so because the packages opened by dart:ffi then remain in memory (what is caused by the running isolates). Then the next time I launch the application, I get errors from them indicating that the libraries have already been opened.

When the isolates are closed, the dart:ffi packages are also closed, and only the plugins associated with the platform remain in memory, so I did not see the errors - that tricked me a bit.

**Comment by kamilklyta** [comment]
Fun fact, there are many packages that pretend to be able to keep isolates alive after closing the application, but they do not work properly. You guys have created one of the most reliable "isolate keep-alive" packages through a bug, or should I say feature 😅

**Comment by fjnoyp** [comment]
Indeed :) ...

We'll look into this and why it's occurring.  Thanks again for sharing this issue.

**Comment by fjnoyp** [comment]
Hi @kamilklyta this is an issue with the underlying Amplify Android Storage library that Amplify Flutter stable is built off of.  We will need to create an issue on their repository and see if they can fix it.  I'll create the issue and link you in later today. 

In the meantime, I'll check internally if there's something else you can do to unblock yourself here.

**Comment by fjnoyp** [comment]
Hi @kamilklyta if you **do not ** need background upload/download support (ie. it's okay that storage s3 upload/download operations are canceled if the user backgrounds the app) then we recommend you switch to using the developer preview version of Amplify Storage:

https://docs.amplify.aws/lib/storage/getting-started/q/platform/flutter/#install-amplify-libraries

We have also created an issue on Amplify Android to get their feedback on what is happening:

https://github.com/aws-amplify/amplify-android/issues/2324

**Comment by kamilklyta** [comment]
Is the background upload feature disabled on the preview version, or is it because another native library backs it?

Is there a way to disable background upload on a native package without switching to the preview version? 🤔 

For this, we will have to migrate all packages related to amplify-flutter 🤔. I will try it in the coming days and let you know the result. Thanks again for your help 💪.

**Comment by fjnoyp** [comment]
The preview version rewrites the entire Amplify functionality to be dart first instead of relying on the Amplify iOS and Android libraries.  So background upload feature simply hasn't been developed yet. 

There is currently no way to disable background upload like that.  

What packages do you use from `amplify-flutter`?

Android team is looking into this too so we can monitor for updates on the corresponding Android issue as well.

**Comment by kamilklyta** [comment]
Currently, I am using `amplify_flutter`, `amplify_auth_cognito`, `amplify_storage_s3`, `amplify_analytics_pinpoint`, and of course the transitive packages like `amplify_core` and platform-dependant interfaces.

**Comment by fjnoyp** [comment]
Thanks for letting us know @kamilklyta.  That should be okay but we can double check before you upgrade packages as well. 

The Amplify Android team is also looking into the issue on their side as well.  We had an update here: 
https://github.com/aws-amplify/amplify-android/issues/2324

**Comment by fjnoyp** [comment]
Hi @kamilklyta I just spoke with the Android team and it sounds like you can disable the background upload/download by modifying the Android Manifest.  We can make that change and test if it fixes your issue.  This way you will not need to update to preview version to fix this issue of yours.

**Comment by fjnoyp** [comment]
Hey @kamilklyta was this able to resolve your issue?

**Comment by kamilklyta** [comment]
I think so 🤔.
The change has passed our local tests.
After the release of the new version of the app, I will confirm if the statistics on firebase have improved.

**Comment by Jordan-Nelson** [comment]
@kamilklyta - Have you had a chance to confirm if this resolved the issue?

**Comment by fjnoyp** [comment]
Hi @kamilklyta I'm going to close this issue for now.  If you have other problems, feel free to reopen or create another issue.
